cumulus/primitives/storage-weight-reclaim/src/lib.rs

```diff
@@ -119,53 +126,51 @@ impl<T: Config + Send + Sync> core::fmt::Debug for StorageWeightReclaim<T> {
 	}
 }
 
-impl<T: Config + Send + Sync> SignedExtension for StorageWeightReclaim<T>
+impl<T: Config + Send + Sync> TransactionExtensionBase for StorageWeightReclaim<T> {
+	const IDENTIFIER: &'static str = "StorageWeightReclaim";
+	type Implicit = ();
+}
+
+impl<T: Config + Send + Sync, Context> TransactionExtension<T::RuntimeCall, Context>
+	for StorageWeightReclaim<T>
 where
 	T::RuntimeCall: Dispatchable<Info = DispatchInfo, PostInfo = PostDispatchInfo>,
 {
-	const IDENTIFIER: &'static str = "StorageWeightReclaim";
-
-	type AccountId = T::AccountId;
-	type Call = T::RuntimeCall;
-	type AdditionalSigned = ();
+	type Val = ();
 	type Pre = Option<u64>;
 
-	fn additional_signed(
-		&self,
-	) -> Result<Self::AdditionalSigned, sp_runtime::transaction_validity::TransactionValidityError>
-	{
-		Ok(())
-	}
-
-	fn pre_dispatch(
+	fn prepare(
 		self,
-		_who: &Self::AccountId,
-		_call: &Self::Call,
-		_info: &sp_runtime::traits::DispatchInfoOf<Self::Call>,
+		_val: Self::Val,
+		_origin: &T::RuntimeOrigin,
+		_call: &T::RuntimeCall,
+		_info: &DispatchInfoOf<T::RuntimeCall>,
 		_len: usize,
-	) -> Result<Self::Pre, sp_runtime::transaction_validity::TransactionValidityError> {
+		_context: &Context,
+	) -> Result<Self::Pre, TransactionValidityError> {
 		Ok(get_proof_size())
```

Comment:
This extension is very tricky. Initially I hoped the reclaim for just the call would be fine for now and we could work on this particular extension later, since it also doesn't have benchmarks yet.

I don't think a 2 extension solution is viable because it's very difficult to pass information from one extension to the other in the pipeline, in this case it's the intial proof size returned by `prepare`  which I think will be different depending on the position and call site (that's probably the point though :grin: ). Maybe we can construct it with the existing information in `DispatchInfo` and `PostDispatchInfo` (or minimal changes to them), but I'm not sure since I haven't looked at this in a while. Is there something I'm missing on this one?

Reply:
2 extensions could use some transient storage to store information.

I think the Extension tries to be more accurate than `DispatchInfo` and `PostDispatchInfo` by using the `ProofSizeExt` runtime extension.

It increases if the proof size as calculated with the runtime extension is more the what is calculated with the dispatch info.
It increases if the proof size as calculated with the runtime extension is less the what is calculated with the dispatch info.

Reply:
How would it look like? Would it keep track of the last value returned as well as the current one (right now it's recording only the current size, maybe we can store checkpoint values)?

Reply:
I meant it would store somewhere in the storage the value: like at the key twox64("some_unique_identifier").

and then at the post_dispatch stage. it would get and remove this value.

So from the point of view of the block execution nothing should leak into the PoV.

If cumulus has a standard pallet we could then used a proper storage in this pallet.

Reply:
I don't think it's possible with the default cumulus pallet, although that would probably be `parachain_system`? I think the better approach would be to create a pallet just for this extension. We could do this with pallet storage and we would have a nice way to stick in a `WeightInfo` type with proper pallet benchmarks.

Reply:
for context, following PR is here https://github.com/paritytech/polkadot-sdk/pull/5234